.github/workflows/web_app_deploy.yaml

```diff
@@ -0,0 +1,159 @@
+        required: false
+        default: true
+
+concurrency:
+  group: ${{ github.workflow }}-cd
+  cancel-in-progress: true
+
+env:
+  BUNDLE_NAME: bundle
+
+jobs:
+  build:
+    name: Build Artifact
+    runs-on: ubuntu-20.04
+    env:
+      WORKSPACE: ${{ inputs.workspace_name }}
+
+    steps:
+      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        name: Checkout
+
+      - name: Prune
+        run: npx turbo@1.13.3 prune --scope ${{ env.WORKSPACE }}
+
+      - name: Setup Node.js
+        uses: actions/setup-node@60edb5dd545a775178f52524783378180af0d1f8 # v4.0.2
+        with:
+          node-version-file: ".node-version"
+          cache: "yarn"
+          cache-dependency-path: "./out/yarn.lock"
+
+      - name: Install dependencies
+        run: yarn install --immutable
+        working-directory: ./out
+
+      - name: Restore turbo cache
+        uses: actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319 # v4.0.1
+        with:
+          path: ./out/node_modules/.cache/turbo
+          key: ${{ runner.os }}-turbo-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-turbo
+
+      - name: Build
+        run: yarn build
+        working-directory: ./out
+
+      - name: Build the artifcat
+        id: make-artifact
+        run: |
+          yarn bundle
+          echo "artifact-path=$(realpath bundle.zip)" >> "$GITHUB_OUTPUT"
+        working-directory: ./out/apps/${{ env.WORKSPACE }}
+
+      - name: Upload Artifact
+        uses: actions/upload-artifact@694cdabd8bdb0f10b2cea11669e1bf5453eed0a6 # v4.2.0
+        with:
+          name: ${{ env.BUNDLE_NAME }}
+          path: ${{ steps.make-artifact.outputs.artifact-path }}
+          if-no-files-found: error
+          retention-days: 7
+
+  deploy:
+    name: Deploy
+    if: ${{ !github.event.act }}
+    needs: [build]
+    runs-on: ${{ inputs.use_private_agent == true && 'self-hosted' || 'ubuntu-20.04' }}
+    environment: ${{ inputs.environment }}-cd
+    permissions:
+      id-token: write
+      contents: read
+    env:
+      ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
+      ARM_SUBSCRIPTION_ID: ${{ secrets.ARM_SUBSCRIPTION_ID }}
+      ARM_TENANT_ID: ${{ secrets.ARM_TENANT_ID }}
+      ARM_USE_OIDC: true
+      RESOURCE_GROUP_NAME: ${{ inputs.resource_group_name }}
+      WEB_APP_NAME: ${{ inputs.web_app_name}}
+      HEALTH_CHECK_PATH: ${{ inputs.health_check_path }}
+      USE_STAGING_SLOT: ${{ inputs.use_staging_slot }}
+
+    steps:
+      - name: Download Artifact
+        uses: actions/download-artifact@6b208ae046db98c579e8a3aa621ab581ff575935 # v4.1.1
+        with:
+          name: ${{ env.BUNDLE_NAME }}
+
+      - name: Azure Login
+        uses: azure/login@v2 # v2.0.0
+        with:
+          client-id: ${{ env.ARM_CLIENT_ID }}
+          tenant-id: ${{ env.ARM_TENANT_ID }}
+          subscription-id: ${{ env.ARM_SUBSCRIPTION_ID }}
+
+      - name: Deploy
+        uses: azure/webapps-deploy@v2
```

Comment:
I did some deployment test with AZ CLI to remove the `curl` ping due to recent issues and applying a more reliable approach.

Here the test:
scenario:
- app with a single endpoint that responds 200
- deploy from local via az cli
- 3 instances per instance

steps:
1. application deployment that responds 200 to the root with the `az webapp deploy --src-path app.zip` command, which seems to be (almost) synchronous
2. staging slot creation:
    - the same application that runs on prod slot responds both on `<app>-staging.azurewebsites.net` and `<app>.azurewebsites.net?x-ms-routing-name=staging`
3. enabled health checks
    - all instances are green on both slots
4. changed the code to reply 400 instead of 200
5. deployment of the “broken” app on staging slot
    - the deployment “never” finishes, in the sense that it tries again until the timeout of about 15 minutes expires (timeout is customizable)
    - as expected, the working app continues to respond 200 to the url `<app>.azurewebsites.net`
    - the broken app responds with 400 to the url `<app>-staging.azurewebsites.net`
    - unexpectedly, the url `<app>.azurewebsites.net?x-ms-routing-name=staging` responds with 200 instead of 400
    - prod shows all healthy instances, staging shows them all red
6. slot swap
    - as expected, `<app>.azurewebsites.net` responds 400 and `<app>-staging.azurewebsites.net` responds 200
    - <app>.azurewebsites.net?x-ms-routing-name=staging responds 400
    - as expected, healths are reversed (prod red, staging green)
7. new slot swap
    - green prod, red staging
8. set of these variables in both slots:
    - WEBSITE_WARMUP_PATH: / (`A relative path on the site that should be pinged whenever the site restarts (not only during slot swaps)
    - WEBSITE_SWAP_WARMUP_PING_STATUSES: 200 (Valid HTTP response codes for the warm-up operation. Add this app setting with a comma-separated list of HTTP codes. An example is 200,202 . If the returned status code isn't in the list, the warmup and swap operations are stopped. By default, all response codes are valid.)
9. attempted new swap, got an error:
   - Failed to swap slots 'staging' and 'production': Cannot swap slots for site 'io-d-itn-test-autoscaler-app-02' because the application initialization in 'staging' slot either took too long or failed. Please check AppInit module configuration or try using swap with preview if application initialization time is very long.
    - as expected, the app in production continues to respond as before

Given this test, I would set these environment variables in all services (including our module update), remove curl from the pipeline and run the swap command.
I would also deploy the package using with az cli because the action [uses the `async=true` query string](https://github.com/microsoft/pipelines-appservice-lib/blob/0af7ef8d67738e5249f8f5e3c5e8a036b940ecf5/packages/appservice-rest/src/Utilities/KuduServiceUtility.ts#L108). Through az cli we could use a synchronous approach and set the timeout to 1 minute or little more.
I would also set the variable `WEBSITE_SWAP_WARMUP_PING_PATH` (`The path to ping over HTTP to warm up your site. Add this app setting by specifying a custom path that begins with a slash as the value`), which from what I understand is the same as WEBSITE_WARMUP_PATH but is only called for swap operations (I think in case you want to use different endpoints for swap and restart, but you might as well set them both)

Reply:
I updated the workflow to use `az webapp deploy` instead the GitHub Action. I think we should update our web apps tf modules to include `WEBSITE_WARMUP_PATH` by default (pointing to the health check path)

Reply:
I like your first change, I'm going to approve the PR on my side
Instead, I don't think adding the `WEBSITE_WARMUP_PATH` variable is correct as we've seen some issues while spinning up the container (probably things are not ready and it is restarted multiple times), while I would add `WEBSITE_SWAP_WARMUP_PING_PATH` with the health ping